tools/sctrace/README.md

```diff
@@ -0,0 +1,357 @@
+sctrace patterns.scml -- ls -la
+sctrace patterns.scml --quiet -- ./my_program arg1 arg2
+```
+
+### Offline Mode (Log file analysis)
+
+Analyze an existing strace log file:
+
+```bash
+sctrace patterns.scml --input trace.log
+sctrace patterns.scml --input trace.log --quiet
+```
+
+### Options
+
+- `--input <FILE>`: Specify input file for offline mode
+- `--quiet`: Enable quiet mode (only show unsupported calls)
+
+## SCML (System Call Matching Language)
+
+SCML is a domain-specific language for defining system call patterns. For more
+detailed information about SCML syntax and features, please refer to the
+[official SCML documentation](https://asterinas.github.io/book/kernel/linux-compatibility/limitations-on-system-calls/system-call-matching-language.html).
+
+It supports:
+
+### Basic Syntax
+
+```scml
+// Basic system call pattern
+read(fd, buf, count);
+
+// Constrained parameters
+open(pathname, flags = O_RDONLY | O_WRONLY, mode);
+
+// Comments (C-style)
+// This is a comment
+```
+
+### Built-in Types
+
+- `<INTEGER>`: Constrains parameter to integer values
+- `<PATH>`: Constrains parameter to file path patterns
+
+### Named Bitflags
+
+Define reusable flag sets:
+
+```scml
+access_mode = O_RDONLY | O_WRONLY | O_RDWR;
+open_flags = O_CREAT | O_EXCL | O_TRUNC;
+
+open(pathname = <PATH>, flags = <access_mode> | <open_flags>, mode);
+```
+
+### Struct Patterns
+
+Define structured data patterns:
+
+```scml
+struct stat = {
+    st_mode = <INTEGER>,
+    st_size = <INTEGER>,
+    ..
+};
+
+stat(pathname = <PATH>, statbuf = <stat>);
+```
+
+### Array Patterns
+
+Define array constraints:
+
+```scml
+poll(fds = [{ fd, events = POLLIN | POLLOUT, .. }], nfds, timeout);
+```
+
+## Examples
+
+### Example 1: Basic File Operations
+
+Create `file_ops.scml`:
+```scml
+openat(dirfd, flags = O_RDONLY | O_WRONLY | O_RDWR, mode);
+read(fd, buf, count = <INTEGER>);
+write(fd, buf, count = <INTEGER>);
+close(fd);
+```
+
+Run:
+```bash
+sctrace file_ops.scml -- cat /etc/passwd
+```
+
+### Example 2: Network Operations
+
+Create `network.scml`:
+```scml
+socket(domain = AF_INET | AF_INET6, type = SOCK_STREAM | SOCK_DGRAM, protocol);
+connect(sockfd, addr, addrlen);
+send(sockfd, buf, len, flags);
+recv(sockfd, buf, len, flags);
+```
+
+Run:
+```bash
+sctrace network.scml -- curl http://example.com
+```
+
+### Example 3: Using Asterinas Compatibility Patterns
+
+Use the provided [asterinas.scml](../../book/src/kernel/linux-compatibility/limitations-on-system-calls/asterinas.scml) (work in progress) and
+test with various commands:
+
+```bash
+# Monitor file system operations
+sctrace asterinas.scml -- tree .
+
+# Monitor process information calls
+sctrace asterinas.scml -- top
+
+# Monitor network operations
+sctrace asterinas.scml -- ping 127.0.0.1
+```
+
+### Example 4: Offline Analysis
+
+```bash
+# Generate trace log
+strace -o trace.log ls -la
+
+# Analyze with sctrace
+sctrace patterns.scml --input trace.log
+```
+
+## strace Output Format Support
```

Comment:
Remove this section. The primary reader of this README is the end users of the tool, not developers. The end users are not interested in learning the format of strace.